ports/stm32/machine_i2s.c

```diff
@@ -0,0 +1,1109 @@
+        // empty
+        return false;
+    }
+
+    *data = rbuf->buffer[rbuf->head];
+    rbuf->head = (rbuf->head + 1) % rbuf->size;
+    return true;
+}
+
+bool ringbuf_is_empty(ring_buf_t *rbuf) {
+    return rbuf->head == rbuf->tail;
+}
+
+bool ringbuf_is_full(ring_buf_t *rbuf) {
+    return ((rbuf->tail + 1) % rbuf->size) == rbuf->head;
+}
+
+size_t ringbuf_available_data(ring_buf_t *rbuf) {
+    return (rbuf->tail - rbuf->head + rbuf->size) % rbuf->size;
+}
+
+size_t ringbuf_available_space(ring_buf_t *rbuf) {
+    return rbuf->size - ringbuf_available_data(rbuf) - 1;
+}
+
+//  For 32-bit audio samples, the STM32 HAL API expects each 32-bit sample to be encoded
+//  in an unusual byte ordering:  Byte_2, Byte_3, Byte_0, Byte_1
+//      where:  Byte_0 is the least significant byte of the 32-bit sample
+//
+//  The following function takes a buffer containing 32-bits sample values formatted as little endian
+//  and performs an in-place modification into the STM32 HAL API convention
+//
+//  Example:
+//
+//   wav_samples[] = [L_0-7,   L_8-15,  L_16-23, L_24-31, R_0-7,   R_8-15,  R_16-23, R_24-31] =  [Left channel, Right channel]
+//   stm_api[] =     [L_16-23, L_24-31, L_0-7,   L_8-15,  R_16-23, R_24-31, R_0-7,   R_8-15] = [Left channel, Right channel]
+//
+//   where:
+//      L_0-7 is the least significant byte of the 32 bit sample in the Left channel
+//      L_24-31 is the most significant byte of the 32 bit sample in the Left channel
+//
+//   wav_samples[] =  [0x99, 0xBB, 0x11, 0x22, 0x44, 0x55, 0xAB, 0x77] = [Left channel, Right channel]
+//   stm_api[] =      [0x11, 0x22, 0x99, 0xBB, 0xAB, 0x77, 0x44, 0x55] = [Left channel, Right channel]
+//
+//   where:
+//      LEFT Channel =  0x99, 0xBB, 0x11, 0x22
+//      RIGHT Channel = 0x44, 0x55, 0xAB, 0x77
+STATIC void reformat_32_bit_samples(int32_t *sample, uint32_t num_samples) {
+    int16_t sample_ms;
+    int16_t sample_ls;
+    for (uint32_t i = 0; i < num_samples; i++) {
+        sample_ls = sample[i] & 0xFFFF;
+        sample_ms = sample[i] >> 16;
+        sample[i] = (sample_ls << 16) + sample_ms;
+    }
+}
+
+STATIC int8_t get_frame_mapping_index(int8_t bits, format_t format) {
+    if (format == MONO) {
+        if (bits == 16) {
+            return 0;
+        } else { // 32 bits
+            return 1;
+        }
+    } else { // STEREO
+        if (bits == 16) {
+            return 2;
+        } else { // 32 bits
+            return 3;
+        }
+    }
+}
+
+STATIC int8_t get_dma_bits(uint16_t mode, int8_t bits) {
+    if (mode == I2S_MODE_MASTER_TX) {
+        if (bits == 16) {
+            return I2S_DATAFORMAT_16B;
+        } else {
+            return I2S_DATAFORMAT_32B;
+        }
+        return bits;
+    } else { // Master Rx
+        // always read 32 bit words for I2S e.g.  I2S MEMS microphones
+        return I2S_DATAFORMAT_32B;
+    }
+}
+
+uint32_t fill_appbuf_from_ringbuf(machine_i2s_obj_t *self, mp_buffer_info_t *appbuf) {
```

Comment:
this can be STATIC

Reply:
fixed in https://github.com/miketeachman/micropython/commit/0f7b59c92e8173d86f596097f9f8fc5ef3e3cfd5